[WritingTools] ObjC exception in `preferredFilenameForElement` under `-[NSFileWrapper setPreferredFilename:]`

whsieh opened the pull request:
#### 41bef39eadb8d5a1a4c4f79eaa39834997d8d2e2
<pre>
[WritingTools] ObjC exception in `preferredFilenameForElement` under `-[NSFileWrapper setPreferredFilename:]`
<a href="https://bugs.webkit.org/show_bug.cgi?id=288771">https://bugs.webkit.org/show_bug.cgi?id=288771</a>
<a href="https://rdar.apple.com/145408164">rdar://145408164</a>

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Make `fileWrapperForElement(const HTMLImageElement&amp;)` robust in the case where the image element is
backed by an attachment element without a `title` attribute. Currently, our logic to extract the
`preferredFilenameForElement` does the following:

```
if (RefPtr attachmentElement = element.attachmentElement())
    return attachmentElement-&gt;attachmentTitle();
```

...which causes us to pass an empty string as the `NSFileWrapper`&apos;s preferred file name, which in
turn causess an ObjC exception to be thrown. We can instead make this more robust, by only returning
the attachment element&apos;s title if it&apos;s non-empty.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(preferredFilenameForElement):

See above. Additionally rearrange this helper, to guarantee that we *always* return a non-empty
string by guarding every early return with an `!….isEmpty()` check. The fallback return value,
`copyImageUnknownFileLabel()`, just looks up a hard-coded fallback string from the localized strings
dictionary, so it should always exist.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingWithUntitledImageAttachment)):

Add a new API test to exercise this crash fix.

Canonical link: <a href="https://commits.webkit.org/291293@main">https://commits.webkit.org/291293@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de21b4708bd8d4e2716a95c443242e727be0d48e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1682 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97514 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43037 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94580 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20533 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70879 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/28330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95532 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83783 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51211 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9067 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1421 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42368 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1362 "Found 1 new test failure: swipe/pushState-back-swipe-verify-ua-transition.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99541 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19581 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19831 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79665 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79167 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/19627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23703 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19565 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19252 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22712 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20992 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2334,32 +2334,40 @@ static NSInteger _colCompare(id block1, id block2, void *)
+        RefPtr frame = element.document().frame();
+        if (frame->loader().frameHasLoaded()) {
+            RefPtr dataSource = frame->loader().documentLoader();
+            if (auto resource = dataSource->subresource(url.get())) {
```
```suggestion
            if (RefPtr resource = dataSource->subresource(url.get())) {
```

whsieh replied:
Fixed — thanks both for the reviews!

whsieh committed:
[WritingTools] ObjC exception in `preferredFilenameForElement` under `-[NSFileWrapper setPreferredFilename:]`
https://bugs.webkit.org/show_bug.cgi?id=288771
rdar://145408164

Reviewed by Aditya Keerthi and Abrar Rahman Protyasha.

Make `fileWrapperForElement(const HTMLImageElement&)` robust in the case where the image element is
backed by an attachment element without a `title` attribute. Currently, our logic to extract the
`preferredFilenameForElement` does the following:

```
if (RefPtr attachmentElement = element.attachmentElement())
    return attachmentElement->attachmentTitle();
```

...which causes us to pass an empty string as the `NSFileWrapper`'s preferred file name, which in
turn causess an ObjC exception to be thrown. We can instead make this more robust, by only returning
the attachment element's title if it's non-empty.

* Source/WebCore/editing/cocoa/HTMLConverter.mm:
(preferredFilenameForElement):

See above. Additionally rearrange this helper, to guarantee that we *always* return a non-empty
string by guarding every early return with an `!….isEmpty()` check. The fallback return value,
`copyImageUnknownFileLabel()`, just looks up a hard-coded fallback string from the localized strings
dictionary, so it should always exist.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(TEST(WritingTools, ProofreadingWithUntitledImageAttachment)):

Add a new API test to exercise this crash fix.

Canonical link: https://commits.webkit.org/291293@main
